### ISSUE TITLE
Improve WktReader

### ISSUE THREAD

**Issue by FObermaier** [NetTopologySuite/NetTopologySuite#190]
The WktReader is currently not capable of reading M- or ZM geometries properly.

To fix that we need to overcome the limitations of the current implmentation that it internally uses `List<Coordinate>` to gather coordinates during read. It then uses an array of `Coordinate`s (`ToArray()`) to create the `ICoordinateSequence` for the geometry.

Even though SRID tokens are not part of the WKT spec, it is used in a wide range of geospatial applications and frameworks, so I think the WKTReader should do that too. It already does that, but the  current implementation falls short.

The 'Z', 'M' and 'ZM' are read alright, but not handled at all afterwards. This needs to be fixed. I propose the following:

- [ ] Pull `CoordinateBuffer` into core `NetTopologySuite` project and use it to gather ordinate values in the WKTReader
- [ ] Fix handling of (possible) SRID values properly
- [ ] Create delegate functions for parsing 2D, 2DM, 3D, 3DM coordinates and assign those based on the parsed ordinate tokens.

**I am aware that this might cause some issues not producing the same input geometries for unit tests throughout the suite.**
- [ ] Possibly create a "ReadCoordinatesTheOldWay" delegate function as fallback.

As the `WKTReader` of JTS has the same limitation, it would be a fine addition to port that for JTS.

**Comment by airbreather** [comment]
Can someone please list on this issue some representative examples of this "extended" WKT that we would like to support?  I'm not committing to add support myself, but if I was going to, then I'd need that to start with.

Ideally one sample for each combination:

| Z | M | SRID |
|-|-|-|
|N|Y|N|
|Y|N|N|
|Y|Y|N|
|N|N|Y|
|N|Y|Y|
|Y|N|Y|
|Y|Y|Y|

**Comment by airbreather** [comment]
> Pull `CoordinateBuffer` into core `NetTopologySuite` project and use it to gather ordinate values in the WKTReader

I've started on something like this (not exactly `CoordinateBuffer`) in branch [`wip/coordinate-list`](https://github.com/NetTopologySuite/NetTopologySuite/tree/wip/coordinate-list).  A major goal of this one was to minimize the storage overhead from `Z` and `M` values when they're not needed... since nearly everything in core JTS / NTS works exclusively on `X` and `Y` values, several consumers (including the application I maintain at work) will want to just store `X` and `Y` in their WKT / WKB specs anyway, so they would not want to pay an extra 16 bytes (or more!) per point just so that other solutions can have their `Z` and/or `M` values.

I know it's not up to the spec we'd need right away (e.g., I started off by having it fail when copying to an `ICoordinateSequence` that demands values for ordinates not stored in the list, but I realize now that we would probably need to special-case `Z` because that's how `CoordinateArraySequence` works...), but I claim that it's a better starting point than `CoordinateBuffer`.

I'm still not committing to take this issue to completion.

**Comment by airbreather** [comment]
So in general, for formal languages like WKT, I'm not a big fan of writing parsers manually (though I've not had much real-world opportunity to refine that opinion since university).  My preference tends to be to use a [parser generator](https://en.wikipedia.org/wiki/Compiler-compiler) that uses a formal specification of the grammar, [such as](https://en.wikipedia.org/wiki/Comparison_of_parser_generators) [ANTLR](https://en.wikipedia.org/wiki/ANTLR), and I've got a [prototype ANTLR grammar](https://github.com/NetTopologySuite/NetTopologySuite/blob/b38b6615ed6467b4ff092c1d9bfcef31314b1cb0/NetTopologySuite/WKT.g4) that seems to work OK.
- Though I think I'd need to re-evaluate that design to facilitate better code reuse in the thing that ultimately processes a parsed instance.
- Doing so would be very complicated, however, because I'd also like to make sure that *the grammar itself* rejects components with different dimension than their container, as the OGC spec requires, instead of lumping that validation in with the code that treats the parse output.

Plus, I'm not sure I appreciate packaging the ANTLR runtime, nor do I feel comfortable slapping a restriction onto anyone who ports what we do here to other languages that it has to work with ANTLR (or any other particular parser generator), so I feel like the best approach here is going to be a hand-rolled parser like what we've got going on right now.

I'm also not sure about the lexer.  At this point in time, given that we're going to have to break from JTS significantly here, I don't appreciate `RToolsUtil.StreamTokenizer` nearly as much... the only credit that I'm willing to give it is that it makes it a lot easier to port code that used `java.io.StreamTokenizer`.  Of course, continuing to use it would mean that we would have a better time getting our changes upstream, so I'd need to be able to make a stronger argument for changing it than what I choose to spend time coming up with right now, which means there's a good chance that it'll continue to be part of this.

With those details in the back of my head, I think I can take this on (I can generate some samples like what I'd identified above), but it would still be quite a bit of a rewrite.
